src/realm/index_string.cpp

```diff
@@ -1111,6 +1171,131 @@ Mixed StringIndex::get(ObjKey key) const
     return m_target_column.get_value(key);
 }
 
+std::set<std::string> StringIndex::tokenize(const StringData text)
+{
+    std::set<std::string> words;
+
+    const char* str = text.data();
+    const size_t len = text.size();
+    size_t start = 0;
+
+    for (size_t i = 0; i < len; ++i) {
+        signed char c = static_cast<signed char>(str[i]); // char may not be signed by default
+
+        // Words are alnum + unicode chars above 128 (special unicode whitespace
+        // chars are not used as word separators)
+        bool is_alnum = (c >= '0' && c <= '9') || (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') ||
+                        c < 0; // sign bit is set = unicode above 128
+
+        if (is_alnum)
+            continue;
+
+        if (i > start) {
+            StringData w(str + start, i - start);
+
+            // All words are converted to lowercase (not very unicode aware)
+            // This is also where we could do stemming.
+            if (auto opt = case_map(w, false)) {
+                words.emplace(*opt);
+            }
+        }
+        start = i + 1;
+    }
+    if (start < len) {
+        StringData w(str + start, len - start);
+        if (auto opt = case_map(w, false)) {
+            words.emplace(*opt);
+        }
+    }
+
+    return words;
+}
+
+void StringIndex::erase(ObjKey key)
+{
+    StringConversionBuffer buffer;
+    StringData value = get(key).get_index_data(buffer);
+    if (m_target_column.is_fulltext()) {
+        std::set<std::string> words = tokenize(value);
+        for (auto& w : words) {
+            erase_string(key, w);
+        }
+    }
+    else {
+        erase_string(key, value);
+    }
+}
+
+void StringIndex::find_all_fulltext(std::vector<ObjKey>& result, StringData value) const
```

Comment:
What do you think about adding a `score` parameter (0-1) so that the results are produced which match or exceed that specified threshold?
1) We can calculate a match score based on how many tokens match instead doing the intersection (which I think may be surprising behaviour)
2) A score based model allows us to improve over time by implementing things like fuzzy matching or prefix/suffix removal without changing the syntax of the query.
3) This would mean that query syntax might change to something like this: `text TEXT(.95) 'object'` or `text TEXT 'object' > 0.95`

Reply:
Sure - this would certainly be an option. Good idea to have the scoring as part of the actual query run. I have a PR later in this series that will add some ranking to the objects. Let's re-visit the question at that point in time.

Reply:
Ok sure. What is your plan with this project? Do you wish to merge this PR into next-major, but not ship it to production as is? I imagine some of the upcoming features will be "breaking" if we had users on this basic version.

Reply:
The idea is to include the file format changes now so that we can gradually add support for this feature without another breaking change. If we at some point finds out that we need to have other information stored in the file we will have a breaking change again, but that is just the nature of things. I don't think we up front can figure out what is eventually needed, so it is better to introduce a very basic version to get some feedback, and that can be done in this way without big costs.

Reply:
Would it be possible to reserve some space for later updates so they are not breaking?

Reply:
The table implementation should be able handle that there is information it does not know how to use, so if is just a matter of adding some additional configuration of the index, then it can be handled.

Reply:
I'd support a "soft" release of this feature. We know users won't begin to use it until we add support for the new index at the object store schema level.